oonib/oonibackend.py

```diff
@@ -102,60 +102,71 @@
+
+class Echo(protocol.Protocol):
+    def dataReceived(self, data):
+        self.transport.write('server echoes ' + data)
+
+def createService(endpoint, role, endpoint_config):
+    if role == 'bouncer':
+        factory = ooniBouncer
+    elif role == 'collector':
+        factory = ooniBackend
+    else:
+        raise Exception("unknown service type")
+
+    service = internet.StreamServerEndpointService(
+        endpoint, factory
+    )
+    service.setName("-".join([endpoint_config['type'], role]))
+    multiService.addService(service)
+    service.startService()
+
 if config.main.tor_hidden_service:
     torconfig = TorConfig()
-    d = startTor(torconfig)
-
-    def getHSEndpoint(data_dir):
-        if LooseVersion(txtorcon_version) >= LooseVersion('0.10.0'):
-            return TCPHiddenServiceEndpoint(reactor,
-                                            torconfig,
-                                            80,
-                                            hidden_service_dir=data_dir)
-        else:
-            return TCPHiddenServiceEndpoint(reactor,
-                                            torconfig,
-                                            80,
-                                            data_dir=data_dir)
-
-    def printOnionEndpoint(endpointService):
-        print ("Exposed %s Tor hidden service on httpo://%s" %
-               (endpointService.name, endpointService.endpoint.onion_uri))
-
-    def addCollector(torControlProtocol):
-        data_dir = os.path.join(torconfig.DataDirectory, 'collector')
-        collector_service = internet.StreamServerEndpointService(
-            getHSEndpoint(data_dir), ooniBackend
-        )
-        collector_service.setName('collector')
-        multiService.addService(collector_service)
-        collector_service.startService()
-        return collector_service
-
-    d.addCallback(addCollector)
-    d.addCallback(printOnionEndpoint)
-
-    if ooniBouncer:
-        def addBouncer(torControlProtocol):
-            data_dir = os.path.join(torconfig.DataDirectory, 'bouncer')
-            bouncer_service = internet.StreamServerEndpointService(
-                getHSEndpoint(data_dir), ooniBouncer
-            )
-            bouncer_service.setName('bouncer')
-            multiService.addService(bouncer_service)
-            bouncer_service.startService()
-            return bouncer_service
-
-        d.addCallback(addBouncer)
-        d.addCallback(printOnionEndpoint)
-else:
-    if ooniBouncer:
-        bouncer_service = internet.TCPServer(8888, ooniBouncer,
-                                             interface="127.0.0.1")
-        multiService.addService(bouncer_service)
-        bouncer_service.startService()
-    collector_service = internet.TCPServer(8889, ooniBackend,
-                                           interface="127.0.0.1")
-    multiService.addService(collector_service)
-    collector_service.startService()
+    configTor(torconfig)
```

Comment:
Yes you are right, should have checked that.

This is actually a very good improvement as I believe it also fixes the issues we were having with priviledge shedding and Tor not starting shown here: https://github.com/TheTorProject/ooni-backend/issues/65